src/query/expression/src/kernels/filter.rs

```diff
@@ -317,4 +311,87 @@ impl Column {
         unsafe { new.set_len(selected) };
         new.into()
     }
+
+    // low-level API using unsafe to improve performance
+    fn filter_string_scalars(values: &StringColumn, filter: &Bitmap) -> StringColumn {
+        debug_assert_eq!(values.len(), filter.len());
+        let selected = filter.len() - filter.unset_bits();
+        if selected == values.len() {
+            return values.clone();
+        }
+        let data = values.data.as_slice();
+        let offsets = values.offsets.as_slice();
+
+        let mut res_offsets = Vec::with_capacity(selected + 1);
+        res_offsets.push(0);
+
+        let mut res_data = vec![];
+        let hint_size = data.len() / (values.len() + 1) * selected;
+
+        static MAX_HINT_SIZE: usize = 1000000000;
+        if hint_size < MAX_HINT_SIZE && values.len() < MAX_HINT_SIZE {
+            res_data.reserve(hint_size)
+        }
```

Comment:
Why limit the hint size?

Reply:
it's hint size, avoid hint too much capacity.